### PR TITLE
Add support for fallback factories in feign client annotation. Fixes …

### PR DESCRIPTION
**Pull request by bpicode**
…issue #1117.1

### CODE REVIEW

File: docs/src/main/asciidoc/spring-cloud-netflix.adoc [hunk 1]
```diff
@@ -1012,6 +1012,29 @@ static class HystrixClientFallback implements HystrixClient {
 }
 ----
 
+If one needs access to the cause that made the fallback trigger, one can use the `fallbackFactory` attribute inside `@FeignClient`.
+
+[source,java,indent=0]
+----
+@FeignClient(name = "hello", fallbackFactory = HystrixClientFallbackFactory.class)
+protected interface HystrixClient {
+    @RequestMapping(method = RequestMethod.GET, value = "/hello")
```
Comment (spencergibb) [inline docs/src/main/asciidoc/spring-cloud-netflix.adoc]: tabs not spaces

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java [hunk 2]
```diff
@@ -29,26 +32,65 @@
...
+		return feign.target(target);
+	}
+
+	private <T> T targetWithFallbackFactory(String feignClientName, FeignContext context,
+											Target.HardCodedTarget<T> target,
+											HystrixFeign.Builder builder,
+											Class<?> fallbackFactoryClass) {
+		FallbackFactory<? extends T> fallbackFactory = (FallbackFactory<? extends T>)
+			getFromContext(feignClientName, context, fallbackFactoryClass, FallbackFactory.class);
+		Object exampleFallback = fallbackFactory.create(new RuntimeException());
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java]: Comment why you are doing this.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java [hunk 3]
```diff
@@ -29,26 +32,65 @@
...
+
+	private <T> T targetWithFallback(String feignClientName, FeignContext context,
+									 Target.HardCodedTarget<T> target,
+									 HystrixFeign.Builder builder, Class<?> fallback) {
+		T fallbackInstance = getFromContext(feignClientName, context, fallback, target.type());
+		return builder.target(target, fallbackInstance);
+	}
+
+	private <T> T getFromContext(String feignClientName, FeignContext context,
+								 Class<?> annotationType, Class<T> targetType) {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java]: `annotationType` doesn't seem like the right variable name. Maybe `beanType`?

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/feign/invalid/FeignClientValidationTests.java [hunk 4]
```diff
@@ -42,10 +43,7 @@
 	@Test
 	public void testNameAndValue() {
 		this.expected.expectMessage("only one is permitted");
-		AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(
-				NameAndValueConfiguration.class);
-		assertNotNull(context.getBean(NameAndValueConfiguration.Client.class));
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/feign/invalid/FeignClientValidationTests.java]: Why did you remove these assertions?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java [hunk 5]
```diff
@@ -29,26 +32,65 @@
...
+				annotationType, feignClientName));
 		}
 
-		if (!target.type().isAssignableFrom(factory.getFallback())) {
+		if (!targetType.isAssignableFrom(annotationType)) {
 			throw new IllegalStateException(
 					String.format(
-							"Incompatible fallback instance. Fallback of type %s is not assignable to %s for feign client %s",
-							factory.getFallback(), target.type(), factory.getName()));
+							"Incompatible fallback/fallbackFactory instance. Fallback/fallbackFactory of type %s is not assignable to %s for feign client %s",
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java]: I'd rather have a specific message. Pass 'fallback' or 'fallbackFactory' into the method.

File: docs/src/main/asciidoc/spring-cloud-netflix.adoc [hunk 6]
```diff
@@ -1012,6 +1012,29 @@ static class HystrixClientFallback implements HystrixClient {
...
+    Hello iFailSometimes();
+}
+
+static class HystrixClientFallbackFactory implements FallbackFactory<HystrixClient> {
```
Comment (spencergibb) [inline docs/src/main/asciidoc/spring-cloud-netflix.adoc]: Shouldn't this be a `@Bean` if I read you code and tests correctly?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java (hunk 2 above)
Comment (bpicode) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java]: Well, I am basically probing the factory to return fallbacks that actually implement the annotated feign interface. The reason for this is that there is no compile-time guarantee for that.
I must admit, that I am not very proud of that line ;)

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/feign/invalid/FeignClientValidationTests.java (hunk 4 above)
Comment (bpicode) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/feign/invalid/FeignClientValidationTests.java]: The exceptions happen at construction time of the AnnotationConfigApplicationContext. The assertions after that were dead code.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java (hunk 2 above)
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java]: Sorry, I was asking to add a comment to the code.
